### ISSUE TITLE
[pimatic-log-reader]  Error: watch ENOENT

### ISSUE THREAD

**Issue by tomansi** [pimatic/pimatic#620]
The message:

error [pimatic]: A uncaught exception occured: Error: watch ENOENT at errnoException (fs.js:1019:11) at FSWatcher.start (fs.js:1051:11) at Object.fs.watch (fs.js:1076:11) at Tail.watch (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:83:32) at null._onTimeout (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:115:22) at Timer.listOnTimeout [as ontimeout](timers.js:110:15) This is most probably a bug in pimatic or in a module, please report it!

The config.json (about pimatic-log-reader):

```
  "devices": [
    {
      "id": "temperature-from-logfile",
      "name": "Temperature",
      "class": "LogWatcher",
      "file": "/usr/local/pimatic-app/iobox/pimatic-setpoints.txt",
      "attributes": [
        {
          "name": "tempset1",
          "type": "number",
          "unit": "°C"
        },
        {
          "name": "tempset2",
          "type": "number",
          "unit": "°C"
        },
        {
          "name": "tempset3",
          "type": "number",
          "unit": "°C"
        }
      ],
      "lines": [
        {
          "match": "tempset1: (.+)",
          "temperature": "$1"
        },
        {
          "match": "tempset2: (.+)",
          "tempset2": "$1"
        },
        {
          "match": "tempset3: (.+)",
          "tempset3": "$1"
        }
      ]
    },
```

other device to show var in front-end:

```
    {
      "id": "idtempset2",
      "name": "Temp. Set 2",
      "class": "VariablesDevice",
      "variables": [
        {
          "name": "tempset2Var",
          "expression": "$temperature-from-logfile.tempset2",
          "type": "number"
        }
      ]
    },
```

and the log file:

```
TEMPERATURE-SETTINGS
tempset1: 11
tempset2: 12
tempset3: 13
zz
```

The variable values into pimatics are:

```
$temperature-from-logfile.tempset1: 0
$temperature-from-logfile.tempset2: 12
$temperature-from-logfile.tempset3: 13
$idtempset2.tempset2Var: 12
```

Obviusly tempset1 is wrong.

I hope that is enough information.

**Comment by leader21** [comment]
try renaming this variable to tempset1

![](http://www.youscreen.de/kwmgalpk52.jpg)

**Comment by tomansi** [comment]
**Sorry, I checked several times but I missed this error.**

I'm a little obsessed with this plugin because I had some problems of abnormal operation.

The variable that reads the last line is not updated correctly unless the last line contains something.

I explain it with an example.

00:00 - Start. The log file contains 3 lines with:
```
tempset1 11
tempset2 12
tempset3: 13
```
and the variables are:
```
$temperature-from-logfile.tempset1: 11
$temperature-from-logfile.tempset2: 12
$temperature-from-logfile.tempset3: 13
```

00:01 - Log file change to 21,22,23 respectively.
          - Variables changes to 21,22,**13**

00:02 - Log file change to 31,32,33.
          - Variables changes to 31,32,**23**

etc.
Thas is: **the variable that reads the value in the last line don't read it properly the first time** (yes a second time, i.e. the variable is read ok when the log file is saved twice even if the values are the same.

However, if the last variable isn't in the bottom line works fine (this is why I put 'zz' in the example above).

Thanks for your time.

PD.- My English is horrible, I hope to explain enough.

**Comment by tomansi** [comment]
Another messages:

17:24:24 error [pimatic-log-reader]: ENOENT, open '/usr/local/pimatic-app/iobox/pimatic-setpoints.txt'

17:20:08 error [pimatic]: A uncaught exception occured: Error: watch ENOENT at errnoException (fs.js:1019:11) at FSWatcher.start (fs.js:1051:11) at Object.fs.watch (fs.js:1076:11) at Tail.watch (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:83:32) at null._onTimeout (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:115:22) at Timer.listOnTimeout [as ontimeout] (timers.js:110:15) This is most probably a bug in pimatic or in a module, please report it!

17:20:07 error [pimatic]: A uncaught exception occured: Error: ENOENT, no such file or directory '/usr/local/pimatic-app/iobox/pimatic-setpoints.txt' at Object.fs.statSync (fs.js:684:18) at Tail.watchEvent (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:98:18) at FSWatcher.<anonymous> (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:84:22) at FSWatcher.EventEmitter.emit (events.js:98:17) at FSEvent.FSWatcher._handle.onchange (fs.js:1039:12) This is most probably a bug in pimatic or in a module, please report it!

03:00:08 error [pimatic]: A uncaught exception occured: Error: watch ENOENT at errnoException (fs.js:1019:11) at FSWatcher.start (fs.js:1051:11) at Object.fs.watch (fs.js:1076:11) at Tail.watch (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:83:32) at null._onTimeout (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:115:22) at Timer.listOnTimeout [as ontimeout] (timers.js:110:15) This is most probably a bug in pimatic or in a module, please report it!

**Comment by sweetpi** [comment]
@tomansi Do you move around / delete the file your are watching?

**Comment by tomansi** [comment]
Hi @sweetpi 

When the log file disappears  (file is deleted or renamed) then the message is:  

```error [pimatic]: A uncaught exception occured: Error: watch ENOENT at errnoException (fs.js:1019:11) at FSWatcher.start (fs.js:1051:11) at Object.fs.watch (fs.js:1076:11) at Tail.watch (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:83:32) at null._onTimeout (/usr/local/pimatic-app/node_modules/pimatic-log-reader/node_modules/tail/tail.js:115:22) at Timer.listOnTimeout [as ontimeout] (timers.js:110:15) This is most probably a bug in pimatic or in a module, please report it!```

and the pluging don't work until pimatic restart.

The other problem I resolve for now adding a dummy last line.

**Comment by mwittig** [comment]
I am closing this now, as there has been no activity on this issue for a long time. The issue has been fixed recently and the fix will released with pimatic-log-reader@0.9.3
